modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RegisteredDomainProcessor.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest.common;
+
+import org.apache.http.conn.util.PublicSuffixMatcher;
+import org.apache.http.conn.util.PublicSuffixMatcherLoader;
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.ConfigurationUtils;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.Map;
+
+public class RegisteredDomainProcessor extends AbstractProcessor {
+    private static final PublicSuffixMatcher SUFFIX_MATCHER = PublicSuffixMatcherLoader.getDefault();
+
+    public static final String TYPE = "registered_domain";
+
+    private final String field;
+    private final String targetField;
+    private final boolean ignoreMissing;
+
+    RegisteredDomainProcessor(
+        String tag,
+        String description,
+        String field,
+        String targetField,
+        boolean ignoreMissing
+    ) {
+        super(tag, description);
+        this.field = field;
+        this.targetField = targetField;
+        this.ignoreMissing = ignoreMissing;
+    }
+
+    public String getField() {
+        return field;
+    }
+
+    public String getTargetField() {
+        return targetField;
+    }
+
+    public boolean getIgnoreMissing() {
+        return ignoreMissing;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        DomainInfo info = getRegisteredDomain(ingestDocument);
+        if (info == null) {
+            if (ignoreMissing) {
+                return ingestDocument;
+            } else {
+                throw new IllegalArgumentException("unable to set domain information for document");
+            }
+        }
+        String domainTarget = targetField + ".domain";
+        String registeredDomainTarget = targetField + ".registered_domain";
+        String subdomainTarget = targetField + ".subdomain";
+        String topLevelDomainTarget = targetField + ".top_level_domain";
```

Comment:
If `target_field` is null or empty, these fields should be added to the root of the document. As written, it creates an empty string object with these as subfields:

```
            "_source" : {
              "" : {
                "subdomain" : "www",
                "registered_domain" : "google.com",
                "top_level_domain" : "com",
                "domain" : "www.google.com"
              }, ...
```